File: api/track.js
function esc(v) {
  // Escape for Telegram HTML parse_mode.
  if (v === undefined || v === null || v === '') return '—';
  return String(v)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
}

function firstIp(xff) {
  if (!xff) return '';
  return String(xff).split(',')[0].trim();
}

module.exports = async function handler(req, res) {
  if (req.method !== 'POST') {
    res.setHeader('Allow', 'POST');
    return res.status(405).json({ error: 'Method not allowed' });
  }

  const token = process.env.TELEGRAM_BOT_TOKEN;
  const chatId = process.env.TELEGRAM_CHAT_ID;
  if (!token || !chatId) {
    return res.status(500).json({ error: 'Telegram env not configured.' });
  }

  // Client fingerprint payload.
  let body = req.body;
  if (typeof body === 'string') {
    try { body = JSON.parse(body); } catch (e) { body = {}; }
  }
  if (!body || typeof body !== 'object') body = {};
  const c = body; // client-collected data

  // -------- Server-side request data --------
  const h = req.headers || {};
  const ip = firstIp(h['x-forwarded-for']) || h['x-real-ip'] || (req.socket && req.socket.remoteAddress) || '';
  const ua = h['user-agent'] || '';
  const lang = h['accept-language'] || '';
  const referer = h['referer'] || h['referrer'] || '';
  const dnt = h['dnt'] || '';

  // Vercel injects geo headers automatically on deployed functions (free).
  let country = h['x-vercel-ip-country'] || '';
  let region = h['x-vercel-ip-country-region'] || '';
  let city = h['x-vercel-ip-city'] || '';
  let lat = h['x-vercel-ip-latitude'] || '';
  let lon = h['x-vercel-ip-longitude'] || '';
  let isp = '';

  // Fallback geo (local dev or missing headers) via free ip-api.com.
  if (!country && ip && !/^(127\.|::1|10\.|192\.168\.|172\.(1[6-9]|2\d|3[01])\.)/.test(ip)) {
    try {
      const r = await fetch('http://ip-api.com/json/' + encodeURIComponent(ip) +
        '?fields=status,country,regionName,city,lat,lon,isp,query');
      if (r.ok) {
        const g = await r.json();
        if (g && g.status === 'success') {
          country = country || g.country;
          region = region || g.regionName;
          city = city || g.city;
          lat = lat || g.lat;
          lon = lon || g.lon;
          isp = g.isp || '';
        }
      }
    } catch (e) { /* geo best-effort */ }
  }

  try { city = decodeURIComponent(city); } catch (e) { /* keep raw */ }
  try { region = decodeURIComponent(region); } catch (e) { /* keep raw */ }

  const mapLink = (lat && lon) ? ('https://maps.google.com/?q=' + lat + ',' + lon) : '';

  // -------- Build Telegram message (HTML) --------
  const L = [];
  L.push('🔔👁️ <b>New Visitor To Website</b>');
  L.push('');
  L.push('🛜 <b>Network</b>');
  L.push('• 🔌 IP: <code>' + esc(ip) + '</code>');
  if (c.webrtcIps) L.push('📡 WebRTC/local IPs: <code>' + esc(c.webrtcIps) + '</code>');
  L.push('📍 Location: ' + esc([city, region, country].filter(Boolean).join(', ')));
  if (isp) L.push('🏢 ISP: ' + esc(isp));
  if (mapLink) L.push('🗺️ Map: <a href="' + esc(mapLink) + '">' + esc(lat + ',' + lon) + '</a>');
  if (c.connection) {
    const n = c.connection;
    L.push('📶 Conn: ' + esc(n.effectiveType) + ' · ' + esc(n.downlink) + 'Mbps' +
      (n.downlinkMax ? '/' + esc(n.downlinkMax) + 'max' : '') + ' · rtt ' + esc(n.rtt) + 'ms' +
      (n.type ? ' · ' + esc(n.type) : '') + (n.saveData ? ' · saveData' : ''));
  }
  if (c.perf) {
    const p = c.perf;
    L.push('⏱️ Timing: ttfb ' + esc(p.ttfb) + 'ms · dns ' + esc(p.dns) + 'ms · tcp ' + esc(p.tcp) + 'ms · dom ' + esc(p.domLoad) + 'ms');
  }

  L.push('');
  L.push('📱 <b>Device / Browser</b>');
  L.push('🧭 UA: <code>' + esc(ua) + '</code>');
  if (c.uaData) L.push('• Hints: ' + esc(c.uaData));
  if (c.uaHighEntropy) {
    const u = c.uaHighEntropy;
    L.push('• UA-full: ' + esc(u.brands || u.uaFullVersion));
    L.push('• Arch: ' + esc(u.arch) + ' ' + esc(u.bitness) + (u.wow64 ? ' wow64' : '') +
      ' · Model: ' + esc(u.model) + ' · OS: ' + esc(u.platformVersion));
  }
  L.push('🏷️ Platform: ' + esc(c.platform) + ' · Vendor: ' + esc(c.vendor) + (c.oscpu ? ' · ' + esc(c.oscpu) : ''));
  L.push('⚙️ CPU cores: ' + esc(c.cores) + ' · RAM: ' + esc(c.ram) + 'GB');
  if (c.jsHeap) L.push('🧠 JS heap: ' + esc(c.jsHeap.used) + '/' + esc(c.jsHeap.limit) + 'MB');
  if (c.storage) L.push('💾 Storage: ' + esc(c.storage.usedMB) + 'MB used / ' + esc(c.storage.quotaGB) + 'GB quota');
  L.push('👆 Touch points: ' + esc(c.touch));
  L.push('🗣️ Languages: ' + esc(c.languages || lang) + ' · Locale: ' + esc(c.locale));
  L.push('🕐 Timezone: ' + esc(c.timezone) + ' (offset ' + esc(c.tzOffset) + ')');
  L.push('🖥️ Screen: ' + esc(c.screen) + ' (avail ' + esc(c.availScreen) + ') · Viewport: ' + esc(c.viewport) + ' · DPR ' + esc(c.dpr));
  if (c.outerWindow) L.push('• Window: ' + esc(c.outerWindow) + ' @ ' + esc(c.screenPos) + (c.orientation ? ' · ' + esc(c.orientation) : ''));
  L.push('🎨 Color depth: ' + esc(c.colorDepth) + '/' + esc(c.pixelDepth) + ' · Dark mode: ' + esc(c.darkMode));
  L.push('🍪 cookies=' + esc(c.cookieEnabled) + ' · DNT=' + esc(c.doNotTrack) + ' · pdf=' + esc(c.pdfViewer) + ' · plugins=' + esc(c.plugins));
  if (c.keyboardLayout) L.push('⌨️ Keyboard: ' + esc(c.keyboardLayout));
  if (c.permissions) L.push('🔐 Permissions: ' + esc(c.permissions));
  if (c.battery) L.push('🔋 Battery: ' + esc(c.battery.level) + '% ' + (c.battery.charging ? '(charging)' : '(on battery)'));

  L.push('');
  L.push('🔮 <b>Fingerprint</b>');
  L.push('🪪 Persistent ID: <code>' + esc(c.persistentId) + '</code>');
  L.push('🎮 GPU: ' + esc(c.gpu) + (c.glVersion ? ' · ' + esc(c.glVersion) : '') +
    (c.glMaxTexture ? ' · maxTex ' + esc(c.glMaxTexture) : '') + (c.glExts ? ' · ' + esc(c.glExts) : ''));
  L.push('🖨️ Canvas: <code>' + esc(c.canvasHash) + '</code> · Audio: <code>' + esc(c.audioHash) + '</code> · Math: <code>' + esc(c.mathHash) + '</code>');
  if (c.fonts) L.push('🔤 Fonts: ' + esc(c.fonts));
  if (c.voices) L.push('🔊 Voices: ' + esc(c.voices));
  if (c.codecs) {
    const cd = c.codecs;
    L.push('🎞️ Codecs: h264=' + esc(cd.h264) + ' av1=' + esc(cd.av1) + ' hevc=' + esc(cd.hevc) + ' webm=' + esc(cd.webm));
  }
  if (c.features) {
    const f = c.features;
    L.push('✨ Display: gamut=' + esc(f.gamut) + ' hdr=' + esc(f.hdr) + ' pointer=' + esc(f.pointer) +
      ' hover=' + esc(f.hover) + ' contrast=' + esc(f.contrast) +
      (f.reducedMotion ? ' reduced-motion' : '') + (f.forcedColors ? ' forced-colors' : '') +
      (f.invertedColors ? ' inverted' : '') + (f.monochrome ? ' monochrome' : ''));
  }
  if (c.webdriver) L.push('🤖 ⚠️ Automation/bot flag: true');

  L.push('');
  L.push('🌍 <b>Visit</b>');
  L.push('📌 Page: ' + esc(c.page) + (c.title ? ' — ' + esc(c.title) : ''));
  L.push('↩️ Referrer: ' + esc(referer || c.referrer));
  if (dnt) L.push('• DNT header: ' + esc(dnt));

  const text = L.join('\n').slice(0, 4000); // Telegram 4096 char cap

  try {
    const tgUrl = 'https://api.telegram.org/bot' + token + '/sendMessage';
    const r = await fetch(tgUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        chat_id: chatId,
        text: text,
        parse_mode: 'HTML',
        disable_web_page_preview: true
      })
    });
    if (!r.ok) {
      const detail = await r.text();
      return res.status(502).json({ error: 'Telegram error', detail: detail.slice(0, 300) });
    }
    return res.status(200).json({ ok: true });
  } catch (err) {
    return res.status(500).json({ error: 'Request failed', detail: String(err).slice(0, 300) });
  }
};

File: api/track.py
import os
import json
import re
import requests
from urllib.parse import quote
from http.server import BaseHTTPRequestHandler


def esc(v):
    if v is None or v == '':
        return '—'
    v_str = str(v)
    return v_str.replace('&', '&amp;').replace('<', '&lt;').replace('>', '&gt;')


def first_ip(xff):
    if not xff:
        return ''
    return str(xff).split(',')[0].strip()


class handler(BaseHTTPRequestHandler):

    def do_POST(self):
        token = os.getenv('TELEGRAM_BOT_TOKEN')
        chat_id = os.getenv('TELEGRAM_CHAT_ID')
        if not token or not chat_id:
            self._respond(500, {'error': 'Telegram env not configured.'})
            return

        content_length = int(self.headers.get('Content-Length', 0))
        raw_body = self.rfile.read(content_length) if content_length > 0 else b'{}'
        try:
            c = json.loads(raw_body)
        except Exception:
            c = {}
        if not isinstance(c, dict):
            c = {}

        h = self.headers
        ip = first_ip(h.get('x-forwarded-for')) or h.get('x-real-ip') or ''
        ua = h.get('user-agent') or ''
        lang = h.get('accept-language') or ''
        referer = h.get('referer') or h.get('referrer') or ''
        dnt = h.get('dnt') or ''

        country = h.get('x-vercel-ip-country') or ''
        region = h.get('x-vercel-ip-country-region') or ''
        city = h.get('x-vercel-ip-city') or ''
        lat = h.get('x-vercel-ip-latitude') or ''
        lon = h.get('x-vercel-ip-longitude') or ''
        isp = ''

        private_ip_pattern = r'^(127\.|::1|10\.|192\.168\.|172\.(1[6-9]|2\d|3[01])\.)'
        if not country and ip and not re.match(private_ip_pattern, ip):
            try:
                geo_url = f'http://ip-api.com/json/{quote(ip)}?fields=status,country,regionName,city,lat,lon,isp,query'
                geo_response = requests.get(geo_url, timeout=5)
                if geo_response.status_code == 200:
                    geo_data = geo_response.json()
                    if geo_data and geo_data.get('status') == 'success':
                        country = country or geo_data.get('country', '')
                        region = region or geo_data.get('regionName', '')
                        city = city or geo_data.get('city', '')
                        lat = lat or str(geo_data.get('lat', ''))
                        lon = lon or str(geo_data.get('lon', ''))
                        isp = geo_data.get('isp', '')
            except Exception:
                pass

        try:
            city_display = quote(city, safe='')
        except Exception:
            city_display = city
        try:
            region_display = quote(region, safe='')
        except Exception:
            region_display = region

        map_link = f'https://maps.google.com/?q={lat},{lon}' if lat and lon else ''

        lines = []
        lines.append('🔔 <b>New Visitor To Website</b>')
        lines.append('')
        lines.append('🌐 <b>Network</b>')
        lines.append('• IP: <code>' + esc(ip) + '</code>')
        if c.get('webrtcIps'):
            lines.append('• WebRTC/local IPs: <code>' + esc(c.get('webrtcIps')) + '</code>')
        location_parts = [city_display, region_display, country]
        location_str = ', '.join(p for p in location_parts if p)
        lines.append('• Location: ' + esc(location_str))
        if isp:
            lines.append('• ISP: ' + esc(isp))
        if map_link:
            lines.append('• Map: <a href="' + esc(map_link) + '">' + esc(f'{lat},{lon}') + '</a>')

        if c.get('connection'):
            n = c.get('connection')
            conn_str = esc(n.get('effectiveType', '')) + ' · ' + esc(n.get('downlink', '')) + 'Mbps'
            if n.get('downlinkMax'):
                conn_str += '/' + esc(n.get('downlinkMax', '')) + 'max'
            conn_str += ' · rtt ' + esc(n.get('rtt', '')) + 'ms'
            if n.get('type'):
                conn_str += ' · ' + esc(n.get('type', ''))
            if n.get('saveData'):
                conn_str += ' · saveData'
            lines.append('• Conn: ' + conn_str)

        if c.get('perf'):
            p = c.get('perf')
            lines.append('• Timing: ttfb ' + esc(p.get('ttfb', '')) + 'ms · dns ' + esc(p.get('dns', '')) +
                         'ms · tcp ' + esc(p.get('tcp', '')) + 'ms · dom ' + esc(p.get('domLoad', '')) + 'ms')

        lines.append('')
        lines.append('💻 <b>Device / Browser</b>')
        lines.append('• UA: <code>' + esc(ua) + '</code>')
        if c.get('uaData'):
            lines.append('• Hints: ' + esc(c.get('uaData')))
        if c.get('uaHighEntropy'):
            u = c.get('uaHighEntropy')
            lines.append('• UA-full: ' + esc(u.get('brands') or u.get('uaFullVersion', '')))
            arch_str = esc(u.get('arch', '')) + ' ' + esc(u.get('bitness', ''))
            if u.get('wow64'):
                arch_str += ' wow64'
            lines.append('• Arch: ' + arch_str + ' · Model: ' + esc(u.get('model', '')) +
                         ' · OS: ' + esc(u.get('platformVersion', '')))

        lines.append('• Platform: ' + esc(c.get('platform', '')) + ' · Vendor: ' +
                     esc(c.get('vendor', '')) + ((' · ' + esc(c.get('oscpu', ''))) if c.get('oscpu') else ''))
        lines.append('• CPU cores: ' + esc(c.get('cores', '')) + ' · RAM: ' + esc(c.get('ram', '')) + 'GB')
        if c.get('jsHeap'):
            lines.append('• JS heap: ' + esc(c.get('jsHeap', {}).get('used', '')) + '/' +
                         esc(c.get('jsHeap', {}).get('limit', '')) + 'MB')
        if c.get('storage'):
            lines.append('• Storage: ' + esc(c.get('storage', {}).get('usedMB', '')) + 'MB used / ' +
                         esc(c.get('storage', {}).get('quotaGB', '')) + 'GB quota')
        lines.append('• Touch points: ' + esc(c.get('touch', '')))
        lines.append('• Languages: ' + esc(c.get('languages') or lang) + ' · Locale: ' + esc(c.get('locale', '')))
        lines.append('• Timezone: ' + esc(c.get('timezone', '')) + ' (offset ' + esc(c.get('tzOffset', '')) + ')')
        lines.append('• Screen: ' + esc(c.get('screen', '')) + ' (avail ' + esc(c.get('availScreen', '')) +
                     ') · Viewport: ' + esc(c.get('viewport', '')) + ' · DPR ' + esc(c.get('dpr', '')))
        if c.get('outerWindow'):
            window_str = '• Window: ' + esc(c.get('outerWindow', '')) + ' @ ' + esc(c.get('screenPos', ''))
            if c.get('orientation'):
                window_str += ' · ' + esc(c.get('orientation', ''))
            lines.append(window_str)

        lines.append('• Color depth: ' + esc(c.get('colorDepth', '')) + '/' + esc(c.get('pixelDepth', '')) +
                     ' · Dark mode: ' + esc(c.get('darkMode', '')))
        lines.append('• cookies=' + esc(c.get('cookieEnabled', '')) + ' · DNT=' + esc(c.get('doNotTrack', '')) +
                     ' · pdf=' + esc(c.get('pdfViewer', '')) + ' · plugins=' + esc(c.get('plugins', '')))
        if c.get('keyboardLayout'):
            lines.append('• Keyboard: ' + esc(c.get('keyboardLayout')))
        if c.get('permissions'):
            lines.append('• Permissions: ' + esc(c.get('permissions')))
        if c.get('battery'):
            b = c.get('battery', {})
            lines.append('• Battery: ' + esc(b.get('level', '')) + '% ' +
                         ('(charging)' if b.get('charging') else '(on battery)'))

        lines.append('')
        lines.append('🧬 <b>Fingerprint</b>')
        lines.append('• Persistent ID: <code>' + esc(c.get('persistentId', '')) + '</code>')
        gpu_str = esc(c.get('gpu', ''))
        if c.get('glVersion'):
            gpu_str += ' · ' + esc(c.get('glVersion', ''))
        if c.get('glMaxTexture'):
            gpu_str += ' · maxTex ' + esc(c.get('glMaxTexture', ''))
        if c.get('glExts'):
            gpu_str += ' · ' + esc(c.get('glExts', ''))
        lines.append('• GPU: ' + gpu_str)
        lines.append('• Canvas: <code>' + esc(c.get('canvasHash', '')) + '</code> · Audio: <code>' +
                     esc(c.get('audioHash', '')) + '</code> · Math: <code>' + esc(c.get('mathHash', '')) + '</code>')
        if c.get('fonts'):
            lines.append('• Fonts: ' + esc(c.get('fonts')))
        if c.get('voices'):
            lines.append('• Voices: ' + esc(c.get('voices')))
        if c.get('codecs'):
            cd = c.get('codecs')
            lines.append('• Codecs: h264=' + esc(cd.get('h264', '')) + ' av1=' + esc(cd.get('av1', '')) +
                         ' hevc=' + esc(cd.get('hevc', '')) + ' webm=' + esc(cd.get('webm', '')))
        if c.get('features'):
            f = c.get('features')
            display_str = ('• Display: gamut=' + esc(f.get('gamut', '')) + ' hdr=' + esc(f.get('hdr', '')) +
                           ' pointer=' + esc(f.get('pointer', '')) + ' hover=' + esc(f.get('hover', '')) +
                           ' contrast=' + esc(f.get('contrast', '')))
            for flag, label in [('reducedMotion', 'reduced-motion'), ('forcedColors', 'forced-colors'),
                                 ('invertedColors', 'inverted'), ('monochrome', 'monochrome')]:
                if f.get(flag):
                    display_str += ' ' + label
            lines.append(display_str)
        if c.get('webdriver'):
            lines.append('• ⚠️ Automation/bot flag: true')

        lines.append('')
        lines.append('📄 <b>Visit</b>')
        lines.append('• Page: ' + esc(c.get('page', '')) + (' — ' + esc(c.get('title', '')) if c.get('title') else ''))
        lines.append('• Referrer: ' + esc(referer or c.get('referrer', '')))
        if dnt:
            lines.append('• DNT header: ' + esc(dnt))

        text = '\n'.join(lines)[:4000]

        try:
            tg_url = f'https://api.telegram.org/bot{token}/sendMessage'
            tg_response = requests.post(tg_url, json={
                'chat_id': chat_id,
                'text': text,
                'parse_mode': 'HTML',
                'disable_web_page_preview': True
            }, timeout=10)
            if tg_response.status_code != 200:
                self._respond(502, {'error': 'Telegram error', 'detail': tg_response.text[:300]})
            else:
                self._respond(200, {'ok': True})
        except Exception as err:
            self._respond(500, {'error': 'Request failed', 'detail': str(err)[:300]})

    def do_OPTIONS(self):
        self.send_response(200)
        self._send_cors_headers()
        self.end_headers()

    def _send_cors_headers(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')

    def _respond(self, status, body):
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self._send_cors_headers()
        self.end_headers()
        self.wfile.write(json.dumps(body).encode())
